### ISSUE TITLE
The code performance

### ISSUE THREAD

**Issue by chehaoa** [facebookresearch/RPG_KDD2025#3]
The code performance is significantly lower than what is reported in the paper, even after multiple runs using the same parameters provided in the code. I apologize for the directness, but may I ask what might be the reason for this discrepancy?
Example:OrderedDict([('ndcg@5', 0.01610289141535759), ('ndcg@10', 0.019428901374340057), ('recall@5', 0.023287825286388397), ('recall@10', 0.03365357592701912)])

**Comment by JiachengLi1995** [comment]
@chehaoa Can we get more details of your experiments?
@hyp1231 Could you please help on this issue? Thanks.

**Comment by chehaoa** [comment]
Thank you very much. I conducted the experiments using a single A100 GPU and strictly followed the reproduction instructions provided in the README, without making any modifications to the code.

The result on the Sports_and_Outdoors dataset is as follows:
Tue 01 Jul 2025 03:05:03 INFO  Test Results: OrderedDict([('ndcg@5', 0.01610289141535759), ('ndcg@10', 0.019428901374340057), ('recall@5', 0.023287825286388397), ('recall@10', 0.03365357592701912)])
The result on the Beauty dataset is as follows:
Mon 30 Jun 2025 13:02:14 INFO  Test Results: OrderedDict([('ndcg@5', 0.035317350178956985), ('ndcg@10', 0.04359418526291847), ('recall@5', 0.05052989348769188), ('recall@10', 0.07610785961151123)])

I truly appreciate your response once again.

**Comment by hyp1231** [comment]
Thank you for raising this issue and for your interest in our work.

After receiving your report, I conducted a full reproduction from scratch. During this process, I discovered a **bug in our model inference code**, which has now been fixed in [this commit](https://github.com/facebookresearch/RPG_KDD2025/pull/4/commits/2898fa4acb6dad0213ff2905c8afc03001e7b132).

### Bug Explanation

Previously, the `num_beams` parameter was directly fetched from the config object rather than being assigned to a member variable `self.num_beams` in the model. In contrast, the other two inference-related parameters, `n_edges` and `propagation_steps`, were correctly stored as member variables.

Due to this inconsistency, our hyperparameter tuning script like the following:

```python
for num_beams in [...]:
    model.num_beams = num_beams
```

had **no actual effect** — the model always used the default `config['num_beams'] = 50`, regardless of the value set in `model.num_beams`.

This has been corrected by explicitly storing `num_beams` as a model attribute and using `self.num_beams` throughout inference.

---

### Reproduction Protocol

To validate the fix, I re-ran the entire pipeline **from scratch**, using the following steps:

1. **Cloned the repo** fresh and verified the latest code.
2. **Downloaded the data** and regenerated all intermediate representations:

   * Used OpenAI API to generate text representations.
   * Used FAISS to compute semantic IDs.
3. **Trained the base model** using the *exact same hyperparameters* as reported in the paper (i.e., best temperature and learning rate), **without additional tuning**.
4. **Re-performed hyperparameter tuning** during the model inference stage over the following ranges:

```python
num_beams ∈ {20, 100, 200}
n_edges ∈ {20, 30, 200, 500}
propagation_steps ∈ {3, 5}
```

---

### Results

#### Recommendation Performance

| Dataset | Metric     | Paper  | Reproduction |
| ------- | ---------- | ------ | ------------ |
| Sports  | Recall\@5  | 0.0314 | 0.0316       |
|         | NDCG\@5    | 0.0216 | 0.0218       |
|         | Recall\@10 | 0.0463 | 0.0469       |
|         | NDCG\@10   | 0.0263 | 0.0267       |
| Beauty  | Recall\@5  | 0.0550 | 0.0546       |
|         | NDCG\@5    | 0.0381 | 0.0376       |
|         | Recall\@10 | 0.0809 | 0.0798       |
|         | NDCG\@10   | 0.0464 | 0.0457       |
| Toys    | Recall\@5  | 0.0592 | 0.0582       |
|         | NDCG\@5    | 0.0401 | 0.0390       |
|         | Recall\@10 | 0.0869 | 0.0870       |
|         | NDCG\@10   | 0.0490 | 0.0483       |
| CDs     | Recall\@5  | 0.0498 | 0.0498       |
|         | NDCG\@5    | 0.0338 | 0.0339       |
|         | Recall\@10 | 0.0735 | 0.0731       |
|         | NDCG\@10   | 0.0415 | 0.0414       |

* Among **16 total metrics**, the reproduction results are *better (or the same) in 7 cases*, *slightly lower in 9*, with **maximum deviation < 0.0011**.
* This confirms the reproducibility of the reported performance after fixing the bug.

---

### Best Inference Hyperparameters & Visited Item Stats

| Dataset | num\_beams | n\_edges | propagation\_steps | #Items | #Visited Items | Visited Ratio |
| ------- | ---------- | -------- | ------------------ | ------ | -------------- | ------------- |
| Sports  | 100        | 30       | 5                  | 18,357 | 4,653.12       | 25.35%        |
| Beauty  | 20         | 200      | 3                  | 12,101 | 6,239.72       | 51.56%        |
| Toys    | 200        | 20       | 3                  | 11,924 | 5,186.54       | 33.14%        |
| CDs     | 20         | 500      | 5                  | 64,443 | 18,897.87      | 29.32%        |

> Compared with the original paper, the ratio of visited items during inference is slightly higher.

---

### Notes on Environment

* My reproduction was conducted on an **NVIDIA A6000 (48GB)**, while the original paper’s experiments were on **RTX 3090 (24GB)**.
* Base model hyperparameters were **not re-tuned**, which may result in slightly suboptimal base performance.

---

### Summary

* The original code had a **bug in model inference** that affected `num_beams`. After fixing, the paper's results are **largely reproducible**.
* **Model conclusions remain valid** — although visited item ratios are slightly higher.
* Apologies for the earlier reproducibility issues, and thank you again for pointing this out. We will also be updating the arXiv version of the paper soon to reflect this fix.

Please feel free to reach out if you have further questions!
